subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/CompilerDaemonClient.java

```diff
@@ -32,7 +35,12 @@ public CompilerDaemonClient(DaemonForkOptions forkOptions, CompilerDaemonWorker
     public <T extends CompileSpec> CompileResult execute(Compiler<T> compiler, T spec) {
         // currently we just allow a single compilation thread at a time (per compiler daemon)
         // one problem to solve when allowing multiple threads is how to deal with memory requirements specified by compile tasks
-        return workerProcess.execute(compiler, spec);
+        BuildOperationWorkerRegistry.Completion workerLease = buildOperationWorkerRegistry.operationStart();
```

Comment:
I think this needs to be in the `CompilerDaemon` returned by `CompilerDaemonManager` (read `WorkerDaemon`/`WorkerDaemonManager` for the public api).  Which is to say, we want this to block before we call `CompilerClientsManager.reserveNewClient()` as this is when the workerProcess gets started.  If we do this here, we will have already started a new process (and consumed at least some of the requisite resources) before we block waiting on a lease.